tools/ci/src/main.rs

```diff
@@ -44,125 +61,295 @@ fn main() {
+        if flags.contains(Flag::KEEP_GOING) {
+            args.push("--keep-going");
+        }
+
+        test_suite.insert(
+            Check::EXAMPLE_CHECK,
+            vec![CITest {
+                command: cmd!(sh, "cargo check {args...}"),
+                failure_message: "Please fix compiler errors for examples in output above.",
+                subdir: None,
+                env_vars: vec![],
+            }],
+        );
     }
 
-    if what_to_run.contains(Check::COMPILE_CHECK) {
+    if checks.contains(Check::COMPILE_CHECK) {
         // Build bevy and check that it compiles
-        cmd!(sh, "cargo check --workspace")
-            .run()
-            .expect("Please fix compiler errors in output above.");
+        let mut args = vec!["--workspace"];
+        if flags.contains(Flag::KEEP_GOING) {
+            args.push("--keep-going");
+        }
+
+        test_suite.insert(
+            Check::COMPILE_CHECK,
+            vec![CITest {
+                command: cmd!(sh, "cargo check {args...}"),
+                failure_message: "Please fix compiler errors in output above.",
+                subdir: None,
+                env_vars: vec![],
+            }],
+        );
     }
 
-    if what_to_run.contains(Check::CFG_CHECK) {
+    if checks.contains(Check::CFG_CHECK) {
         // Check cfg and imports
-        std::env::set_var("RUSTFLAGS", "-D warnings");
-        cmd!(sh, "cargo +nightly check -Zcheck-cfg --workspace")
-            .run()
-            .expect("Please fix failing cfg checks in output above.");
+        let mut args = vec!["-Zcheck-cfg", "--workspace"];
+        if flags.contains(Flag::KEEP_GOING) {
+            args.push("--keep-going");
+        }
+
+        test_suite.insert(
+            Check::CFG_CHECK,
+            vec![CITest {
+                command: cmd!(sh, "cargo +nightly check {args...}"),
+                failure_message: "Please fix failing cfg checks in output above.",
+                subdir: None,
+                env_vars: vec![("RUSTFLAGS", "-D warnings")],
+            }],
+        );
+    }
+
+    // Actually run the tests:
+
+    let mut failed_checks: Check = Check::empty();
+    let mut failure_message: String = String::new();
+
+    // In KEEP_GOING-mode, we save all errors until the end; otherwise, we just
+    // panic with the given message for test failure.
+    fn fail(
+        current_check: Check,
+        failure_message: &'static str,
+        failed_checks: &mut Check,
+        existing_fail_message: &mut String,
+        flags: &Flag,
+    ) {
+        if flags.contains(Flag::KEEP_GOING) {
+            failed_checks.insert(current_check);
+            if !existing_fail_message.is_empty() {
+                existing_fail_message.push('\n');
+            }
+            existing_fail_message.push_str(failure_message);
+        } else {
+            panic!("{failure_message}");
+        }
+    }
+
+    for (check, battery) in test_suite.into_iter() {
```

Comment:
```suggestion
    for (check, battery) in test_suite {
```

`into_iter()` is implied by the `for`-loop :)

Reply:
Hahaha, I have a bad habit of using `.iter()` and then changing it to `.into_iter()` when I actually want move semantics. I'm actually surprised the lints don't catch this!